### ISSUE TITLE
[4.0] Assignment menu function not accessible to blind users

### ISSUE THREAD

**Issue by zwiastunsw** [joomla/joomla-cms#23911]
The user of the screen reader gets access to almost all implemented operating possibilities, but they are not understandable.  
1. The "Menu selection" label is not announced by the screen reader. 
2. A widget to select / expand menu tree is not accessible because
  a. Does not have an accessible name (headings or label)
  b. “Select” & “Expand” labels are not announced by the screen reader
  c. “All”, “None” options are links instead of buttons.
  d. “All link” & “None link” messages announced by the screen reader are incomprehensible to the blind person. 
![select-expand_0](https://user-images.githubusercontent.com/29527200/52651662-ddcefc00-2eec-11e9-94ad-7df7b9a3ade1.png)
![select-expand](https://user-images.githubusercontent.com/29527200/52651677-e3c4dd00-2eec-11e9-9d51-8fc4d002348a.png)

3. The names of the menus in which the user is to select are not announced by the screen reader.
![image](https://user-images.githubusercontent.com/29527200/52651928-62217f00-2eed-11e9-9e35-3ba3beab3fa1.png)

4. Aliases next to menu item names are unnecessary. They do not provide new information, but they cause information noise, confusion, etc. 
![image](https://user-images.githubusercontent.com/29527200/52651951-6f3e6e00-2eed-11e9-8b6e-93f96ac14b0b.png)

5. Accessibility errors in the context menu:
  a. Nonconsecutive level heading used. The menu contains an `<h5>` tag directly following an `<h3>` (“Sub-items”). In order to maintain a consistent outline of the page for assistive technologies, reduce the gap in the heading level by upgrading this tag to an `<h3>` or `<h2>` or see paragraph b. below.
![image](https://user-images.githubusercontent.com/29527200/52652130-ce9c7e00-2eed-11e9-8c4f-2ee85e214f5f.png)

  b. The names of items in the context menu are unclear. They should rather be: “Select sub-items”, “Deselect sub-items”, “Expand sub-items”, “Collapse subitem”. In this case the heading H5 (Sub-items) is not needed. 
  c. When you select or deselect from the context menu, all items in the level are always selected, not just the selected item. Blind screen reader user does not see or hear it.
  d. Icons next to the names of items in the context menu are unnecessary.

6. Search field not accessible
There is a search box on the Assignment Menu tab. The field does not have an accessible label. In addition, it is not keyboard focusable, because it has been excluded from the tabulation order using the `tabindex="-1"` attribute.

**Comment by dgrammatiko** [comment]
@zwiastunsw if you can come up with an HTML markup I'll do the js part...

PS nevermind here is the correct HTML markup:
```
			<div class="card-header">
				<span class="small"><?php echo Text::_('JSELECT'); ?>:
					<button id="treeCheckAll" class="btn btn-link" type="button"><?php echo Text::_('JALL'); ?></button>,
					<button id="treeUncheckAll" class="btn btn-link" type="button"><?php echo Text::_('JNONE'); ?></button>
				</span>
				<span class="width-20">|</span>
				<span class="small"><?php echo Text::_('COM_MODULES_EXPAND'); ?>:
					<button id="treeExpandAll" class="btn btn-link" type="button"><?php echo Text::_('JALL'); ?></button>,
					<button id="treeCollapseAll" class="btn btn-link" type="button"><?php echo Text::_('JNONE'); ?></button>
				</span>
				<input type="text" id="treeselectfilter" name="treeselectfilter" class="form-control search-query float-right" size="16"
					autocomplete="off" placeholder="<?php echo Text::_('JSEARCH_FILTER'); ?>" aria-invalid="false" tabindex="-1">
			</div>
```

Goes to `administrator/components/com_modules/tmpl/module/edit_assignment.php` lines 40-52

**Comment by zwiastunsw** [comment]
@dgrammatiko : Do you want to improve only the switch?
Your code seems to be correct, but I think you first need to rebuild the code responsible for selecting a menu item to define the switch's operations well.

**Comment by dgrammatiko** [comment]
@zwiastunsw ok define the HTML here and I'll do the coding part

**Comment by jwaisner** [comment]
@chmst is this something you can look into for Accessibility?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23911">issues.joomla.org/tracker/joomla-cms/23911</a>.</sub>

**Comment by zwiastunsw** [comment]
@jwaisner: Yes, this page component needs to be rebuilt because it has many accessibility bariers. You can improve particular elements (for ex. select button and expand button), but in my opinion it is better to build it from scratch.

**Comment by brianteeman** [comment]
simply not true but keep repeating it and magically someone will read your mind to decide what the
 problem is and how it can be fixed to satisfy you. until then i guess this will be left as it is because no one will waste their time - I know I wont be making the same mistake again. Writing code and just getting a fail without any explanation or even suggestion taught me my lesson.

**Comment by zwiastunsw** [comment]
If you have nothing to say, you better not say anything. You could finally learn that.

**Comment by brianteeman** [comment]
Words are cheap. Actions have real value.

**Comment by jwaisner** [comment]
Alright, enough. This can be worked we just need someone with the knowledge/time to do so.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23911">issues.joomla.org/tracker/joomla-cms/23911</a>.</sub>

**Comment by chmst** [comment]
I'll re-activate the PRs tomorrow and see what happens. I am highly interested in a accessible backend.

**Comment by hans2103** [comment]
This issue describes 6 issues:

[x] 1 => solved in https://github.com/joomla/joomla-cms/pull/28726
[x] 2 => solved in https://github.com/joomla/joomla-cms/pull/28726
[ ] 3 =>
[x] 4 => solved in https://github.com/joomla/joomla-cms/pull/28944
[ ] 5 => 
[x] 6 => solved in https://github.com/joomla/joomla-cms/pull/28726

**Comment by brianteeman** [comment]
issue 2 was not solved https://github.com/joomla/joomla-cms/pull/28726#issuecomment-722535874
